scripts/build/check_init_priorities.py

```diff
@@ -0,0 +1,316 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Checks the initialization priorities
+
+This script parses the object files in the specified build directory, creates a
+list of known devices and their effective initialization priorities and
+compares that with the device dependencies inferred from the devicetree
+hierarchy.
+
+This can be used to detect devices that are initialized in the incorrect order,
+but also devices that are initialized at the same priority but depends on each
+other, which can potentially break if the linking order is changed.
+"""
+
+import argparse
+import logging
+import os
+import pathlib
+import pickle
+import sys
+
+from elftools.elf.elffile import ELFFile
+from elftools.elf.relocation import RelocationSection
+from elftools.elf.sections import SymbolTableSection
+
+# This is needed to load edt.pickle files.
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..",
+                                "dts", "python-devicetree", "src"))
+from devicetree import edtlib  # pylint: disable=unused-import
+
+# Prefix used for relocation sections containing initialization data, as in
+# sequence of "struct init_entry".
+_INIT_SECTION_PREFIX = (".rel.z_init_", ".rela.z_init_")
+
+# Prefix used for "struct device" reference initialized based on devicetree
+# entries with a known ordinal.
+_DEVICE_ORD_PREFIX = "__device_dts_ord_"
+
+# File name suffix for object files to be scanned.
+_OBJ_FILE_SUFFIX = ".c.obj"
+
+# Defined init level in order of priority.
+_DEVICE_INIT_LEVELS = ["EARLY", "PRE_KERNEL_1", "PRE_KERNEL_2", "POST_KERNEL",
+                      "APPLICATION", "SMP"]
+
+# File name to check for detecting and skiping nested build directories.
+_BUILD_DIR_DETECT_FILE = "CMakeCache.txt"
+
+# List of compatibles for node where the initialization priority should be the
+# opposite of the device tree inferred dependency.
+_INVERTED_PRIORITY_COMPATIBLES = set()
```

Comment:
is this really a constant empty set like the `ALL_CAPS` name implies? if so, you should probably use the `frozenset()` type to make this extremely clear.